### DISCUSSION TITLE
Test framework for lowdefy app

### DISCUSSION THREAD

**Discussion by 13316090972** [lowdefy/lowdefy#1185] [Questions and Help]
Lowdefy is a convinent tool to build web application connectting to multiple data source such as data bases, APIs and apps.
Is there any test framework for the lowdefy app which do not change the code of the lowdefy app under testing?
It's somehow necessary when the data source is under constrution.

**Comment by 13316090972** [comment]
For example, I write a page with 3 steps:
1. fetch data with requests and actions
2. manipulate data with actions and operators
3. display data with blocks
Sometimes the datasource is unstable,therefore step 1 is hanged up. How can i continue my work in step 2 and step 3 then?

**Reply by Gervwyk** [reply]
Just to clarify here, is this something that happens to the production data source? as in what to do if a request fails, or is it simple that you'd like to implement a unit test like framework for your lowdefy config?

**Reply by 13316090972** [reply]
It sounds good. when the datasource is ok, we switch the lowdefy config to sending the normal request , when it goes wrong ,we switch to setting a state  with fake data and bypass the request.

**Reply by Gervwyk** [reply]
To do this you can use the [`try, catch`](https://docs.lowdefy.com/events-and-actions) options under the `Catching Errors` section when you call your request actions. 

Also, see this answer https://github.com/lowdefy/lowdefy/discussions/1163#discussioncomment-2652344
